keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a resource was allocated so that it is only available on
+one node and the Pod cannot run there because other resources like RAM or CPU
+are exhausted on that node, then the Pod cannot run elsewhere. The same applies
+to resources that are available on a certain subset of the nodes and those
+nodes are busy.
+
+Delayed allocation solves this by integrating allocation with Pod scheduling:
+an attempt to schedule a Pod triggers allocation of pending resources for nodes
+that the scheduler has deemed suitable. Scheduling the pod is then put on hold
+until all resources are allocated. This avoids scenarios where a Pod is
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+### Sharing a single ResourceClaim
+
+Pods reference resource claims in a new PodSpec.ResourceClaims list. Each
+resource in that list can then be made available to one or more containers in
+that Pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time or an unlimited number of pods. Support for additional
+constraints (maximum number of pods, maximum number of nodes) could be
+added once there are use cases for those.
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### Ephemeral vs. persistent ResourceClaims lifecycle
+
+A persistent ResourceClaim has a lifecyle that is independent of any particular
+pod. It gets created and deleted by the user. This is useful for resources
+which are expensive to configure and that can be used multiple times by pods,
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
```

Comment:
This RBAC question also came up when generic ephemeral inline volumes were introduced. That situation was even worse because real clusters might have had PVC creation disabled. We "solved" it by carefully warning about the security impact of having the new feature enabled.

This is also what we need to do here (will add the warning and more explanations): RBAC rules cannot be used to prevent users from creating ResourceClaims. If a quota system is desired, then it needs to be implemented by the resource driver via a custom API. It then needs to refuse to allocate a resource for a user when the quota is exhausted.

It has to be done that way because Kubernetes has no concept of "how much" a specific claim might count against such a quota.

Reply:
See the new "User permissions and quotas" subsection under "Risks and Mitigations".

Reply:
I commented elsewhere on quota, too, resolving this,.